### DISCUSSION TITLE
[QUESTION] Safe node shutdown, disk eviction, replicas and orphaned volumes

### DISCUSSION THREAD

**Discussion by aureq** [longhorn/longhorn#8593] [General]
## Question

(I'm raising an issue as it would be great to get the answer included in the official documentation)

From time to time, nodes need to be maintained and restarted, leading to disruption on the k8s side.

If someone wants to do this safely, they should cordon and drain the node according to the [maintenance guide](https://longhorn.io/docs/1.4.3/volumes-and-nodes/maintenance/). As I also understand it, either the node or the disks should be evicted before stopping or restarting the node according to the [documentation](https://longhorn.io/docs/1.4.3/volumes-and-nodes/disks-or-nodes-eviction/).

As most things in k8s are CLI driven (and that not everyone chose to enable various web UI), it would be great to have some CLI commands added to the 2 docs above as they would provide greater clarity to the users.

### Draining a node

For the node drain, I use this command
```
kubectl drain k3s-2 --ignore-daemonsets --delete-emptydir-data --pod-selector='app!=csi-attacher,app!=csi-provisioner,app!=longhorn-admission-webhook,app!=longhorn-conversion-webhook,app!=longhorn-driver-deployer'
```

This was documented ([1](https://longhorn.io/docs/1.4.0/volumes-and-nodes/maintenance/#updating-the-node-os-or-container-runtime), [2](https://longhorn.io/docs/1.4.1/volumes-and-nodes/maintenance/#updating-the-node-os-or-container-runtime)) but [it's not anymore](https://longhorn.io/docs/1.4.2/volumes-and-nodes/maintenance/#updating-the-node-os-or-container-runtime).

Question 1: Should the `drain` command simply be like? And would it be worth mentioning it in the docs that there has been a change introduced in 1.4.2?

```
kubectl drain k3s-2 --ignore-daemonsets
```

I'm asking as we need to perform maintenance from time to time and noticed that we would get orphans volumes after the reboot on the node that was maintained. So far, all volumes have always recovered but it would be best to do it correctly.

### Evicting a node or a disk

For the node/disk eviction, this is a little more unclear as there's no CLI examples on how to do this without the web UI. This means, it's not possible for us to automate this process. Would be great to get this documented as well if possible.

## Environment

 - Longhorn version: 1.4.2
 - Kubernetes version: 1.27.4
 - Node config
   - OS type and version: Debian 12
   - Kernel version: 6.1.52-1
   - CPU per node: 4/8
   - Memory per node: 32GiB
   - Disk type: nvme pcie gen 3
   - Network bandwidth and latency between the nodes: negotiated 1000baseT-FD
 - Underlying Infrastructure: Baremetal

**Comment by derekbit** [comment]
@aureq 
The document https://www.suse.com/support/kb/doc/?id=000021085 might be helpful for your questions.
You can refer to simplified commands in the steps https://github.com/longhorn/longhorn/issues/3304#issuecomment-1467174481.
 
We will add commands in the official document.

**Comment by derekbit** [comment]
> For the node/disk eviction, this is a little more unclear as there's no CLI examples on how to do this without the web UI. This means, it's not possible for us to automate this process. Would be great to get this documented as well if possible.

You can evict a disk or a node by updating node.longhorn.io resource, via restful api or by a wrapped [python client](https://longhorn.io/docs/1.4.3/references/longhorn-client-python/).

**Reply by gcarrarom** [reply]
Essentially what I've done here: https://github.com/fancy-whale/KuredHorn
Instead of using longhorn.py, just using the kubernetes python library and adding an annotation not to lose track of what is going on.

**Comment by aureq** [comment]
Thanks a lot @derekbit for the guidance here.

Great to see the command has been simplified. It makes things a lot easier. I'll update my own doc too then.

On the Python side, I wasn't aware. It doesn't appear the library has been published on pypi.org. Is there a tracking issue for this?
It would be a lot easier for anyone to build their own client this way rather than having to copy `longhorn.py`.

Looking to evict a node, how this would work? Simply making the node unschedulable?

I'll investigate during the next few days and I'll update this issue.

**Comment by aureq** [comment]
Hi Derek,

I've looked into this further and there are things worth considering here.

First, the `longhorn.py` isn't an official pypi.org package and so anything developed would require manually maintaining the dependency manually (instead of a versioned package in a public registry).

Second, in order to access the longhorn service, either I would need to do it from the inside of that cluster (not always practical) or expose the service outside (which I'd rather avoid).

Given the 2 points above, I would very welcome feature requests. The first point should be quite straight forward. For the second, could it be possible to consider something like what cert-manager and velero are doing? Afaict, they have a dedicated CLI utility to more easily interact with their service (`cmctl` for cert-manager, and `velero` for Velero).

Let me know what you think.

Thanks

**Comment by c3y1huang** [comment]
https://github.com/longhorn/longhorn/issues/8799
